frontend/heir/mlir/mlir_decorator.py

```diff
@@ -0,0 +1,116 @@
+    # regardless of whether there is another reference
+    _recent = collections.deque(maxlen=config.FUNCTION_CACHE_SIZE)
+
+    __uuid = None
+
+    def __init__(self, name, defn, prefer_literal=False, **kwargs):
+        self._ctor_kwargs = kwargs
+        self._name = name
+        self._defn = defn
+        self._prefer_literal = prefer_literal
+        functools.update_wrapper(self, defn)
+
+    @property
+    def _uuid(self):
+        """
+        An instance-specific UUID, to avoid multiple deserializations of
+        a given instance.
+
+        Note this is lazily-generated, for performance reasons.
+        """
+        u = self.__uuid
+        if u is None:
+            u = str(uuid.uuid1())
+            self._set_uuid(u)
+        return u
+
+    def _set_uuid(self, u):
+        assert self.__uuid is None
+        self.__uuid = u
+        self._memo[u] = self
+        self._recent.append(self)
+
+    def _register(self):
+        # _ctor_kwargs
+        from numba.core.typing.templates import (make_intrinsic_template,
+                                                 infer_global)
+
+        template = make_intrinsic_template(self, self._defn, self._name,
+                                           prefer_literal=self._prefer_literal,
+                                           kwargs=self._ctor_kwargs)
+        infer(template)
+        infer_global(self, types.Function(template))
+
+    def __call__(self, *args, **kwargs):
+        """
+        Calls the Python Impl
+        """
+        _, impl =  self._defn(None, *args, **kwargs)
+        return impl(*args, **kwargs)
+
+    def __repr__(self):
+        return "<intrinsic {0}>".format(self._name)
+
+    def __deepcopy__(self, memo):
+        # NOTE: Intrinsic are immutable and we don't need to copy.
+        #       This is triggered from deepcopy of statements.
+        return self
+
+    def _reduce_states(self):
+        """
+        NOTE: part of ReduceMixin protocol
+        """
+        return dict(uuid=self._uuid, name=self._name, defn=self._defn)
+
+    @classmethod
+    def _rebuild(cls, uuid, name, defn):
+        """
+        NOTE: part of ReduceMixin protocol
+        """
+        try:
+            return cls._memo[uuid]
+        except KeyError:
+            llc = cls(name=name, defn=defn)
+            llc._register()
+            llc._set_uuid(uuid)
+            return llc
+
+
+def mlir(*args, **kwargs):
+    """
+    TODO (#1162): update this doc
+    """
+    # Make inner function for the actual work
+    def _mlir(func):
+        name = getattr(func, '__name__', str(func))
+        llc = _MLIR(name, func, **kwargs)
+        llc._register()
+        return llc
+
+    if not kwargs:
+        # No option is given
+        return _mlir(*args)
+    else:
+        # options are given, create a new callable to recv the
+        # definition function
+        def wrapper(func):
+            return _mlir(func)
+        return wrapper
```

Comment:
```suggestion
    name = getattr(func, '__name__', str(func))
    llc = _MLIR(name, func, **kwargs)
    llc._register()
    return llc
```

Since the only usage we have doesn't need the wrapper, let's leave the rest of this stuff out.

Also, what is `llc`? Looks like copypaste, and I can't figure out what it is supposed to mean from the upstream numba source. Weird that it conflicts with `llc` the llvm compiler, but maybe `c` is for cache?

Reply:
I'm generally not yet happy with how the "intrinsic" thing works, as I've struggled to get something that has (a) produces a nice func call in the bytecode (b) executes normally in pure python and (c) works with type inference.

SInce this is only for matmul, which I think makes sense to drop from the PR, I'll just remove this, too.